src/transactions/ChangeTrustOpFrame.cpp

```diff
@@ -17,6 +17,115 @@
 namespace stellar
 {
 
+static void
+decrementPoolUseCount(AbstractLedgerTxn& ltx, Asset const& asset,
+                      AccountID const& accountID)
+{
+    if (!isIssuer(accountID, asset) && asset.type() != ASSET_TYPE_NATIVE)
+    {
+        auto assetTrustLine = ltx.load(trustlineKey(accountID, asset));
+        if (!assetTrustLine)
+        {
+            throw std::runtime_error("asset trustline is missing");
+        }
+
+        --getTrustLineEntryExtensionV2(
+              assetTrustLine.current().data.trustLine())
+              .liquidityPoolUseCount;
+    }
+}
+
+static void
+managePoolOnDeletedTrustLine(
+    AbstractLedgerTxn& ltx,
+    LiquidityPoolConstantProductParameters const& cpParams,
+    PoolID const& poolID, AccountID const& sourceAccount)
+{
+    decrementPoolUseCount(ltx, cpParams.assetA, sourceAccount);
+    decrementPoolUseCount(ltx, cpParams.assetB, sourceAccount);
+
+    auto poolLtxEntry = loadLiquidityPool(ltx, poolID);
+    if (!poolLtxEntry)
+    {
+        throw std::runtime_error("liquidity pool is missing");
+    }
+
+    auto poolTLCount = --poolLtxEntry.current()
+                             .data.liquidityPool()
+                             .body.constantProduct()
+                             .poolSharesTrustLineCount;
+    if (poolTLCount == 0)
+    {
+        poolLtxEntry.erase();
+    }
+}
+
+bool
+ChangeTrustOpFrame::tryIncrementPoolUseCount(AbstractLedgerTxn& ltx,
+                                             Asset const& asset)
+{
+    if (!isIssuer(getSourceID(), asset) && asset.type() != ASSET_TYPE_NATIVE)
+    {
+        auto assetTrustLine = ltx.load(trustlineKey(getSourceID(), asset));
+        if (!assetTrustLine)
+        {
+            innerResult().code(CHANGE_TRUST_TRUST_LINE_MISSING);
+            return false;
+        }
+
+        if (!isAuthorizedToMaintainLiabilities(assetTrustLine))
+        {
+            innerResult().code(CHANGE_TRUST_NOT_AUTH_MAINTAIN_LIABILITIES);
+            return false;
+        }
+
+        ++prepareTrustLineEntryExtensionV2(
```

Comment:
Probably should do a sanity check here that `liquidityPoolUseCount` isn't `INT64_MAX` before the increment. It should never happen for obvious reasons, but it's UB to overflow a signed integer so we should be cautious.

Reply:
Added a `INT32_MAX ` check for `liquidityPoolUseCount` and a `INT64_MAX ` for `poolSharesTrustLineCount`.